application/common/models/User.php

```diff
@@ -189,17 +194,45 @@ public function rules(): array
     private function validatePassword(): Closure
     {
         return function ($attributeName) {
-            
+
             if ( ! $this->hasPasswordAlready()) {
                 $this->attemptPasswordMigration();
             }
-            
+
             if (! password_verify($this->password, $this->password_hash)) {
                 $this->addError($attributeName, 'Incorrect password.');
             }
         };
     }
 
+    private function validateExpiration(): Closure
+    {
+        return function ($attributeName) {
+            $now = time();
+            $expiration = $this->addGracePeriod($this->getPasswordExpiration());
+
+            if ($now > $expiration) {
+                $this->addError($attributeName, 'Expired password.');
+            }
+        };
+    }
+
+    private function getPasswordExpiration(): string
+    {
+        /** @var $mostRecentPassword PasswordHistory */
+        $mostRecentPassword = $this->getPasswordHistories()
+                                   ->orderBy(['id' => SORT_DESC])
+                                   ->one();
+
+        return $mostRecentPassword->expires();
+
+    }
+
+    private function addGracePeriod(string $expiration): int
```

Comment:
Can we confirm with Steve that this method of adding the grace period will still allow the IdP to force redirection to change password rather than logging in? The idea is 14 days before expiration it will give a warning. After expiration, up to the grace period after logging in they will be prompted/forced to go change password before logging in further. I'm assuming looking at this that it just extends their password expiration by 30 days to allow successful authentication but I'm not sure if the IdP will be aware that it is "expired".